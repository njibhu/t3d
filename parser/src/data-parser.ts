import { BaseType, DataType } from "./types";

interface Definition {
  definitions?: {
    [definition: string]: { [key: string]: DataType | string };
  };
  root: {
    [key: string]: DataType | string;
  };
}

interface ParseFunctionReturn {
  newPosition: number;
  data: any;
}

export class DataParser implements Definition {
  public readonly definitions: Definition["definitions"];
  public readonly root: Definition["root"];
  public DEBUG: boolean;

  constructor(definition: Definition) {
    Object.assign(this, definition);
    this.DEBUG = false;
  }

  public parse(dv: DataView, pos: number): ParseFunctionReturn {
    let position = pos;
    const parsedObject = {};
    for (const key in this.root) {
      const value = this.root[key];
      let parsedResult: ParseFunctionReturn;
      if (typeof value === "string") {
        if (this.DEBUG) console.log(key, position.toString(16), "(parseType)");
        parsedResult = this.parseType(dv, position, value);
      } else {
        const { baseType, subType, length } = value;
        if (this.DEBUG) console.log(key, position.toString(16), baseType, subType, length);
        parsedResult = this[baseType](dv, position, subType, length);
      }
      parsedObject[key] = parsedResult.data;
      position = parsedResult.newPosition;
    }

    return {
      newPosition: position,
      data: parsedObject,
    };
  }

  public parseType(dv: DataView, pos: number, typeDefinitionName: string): ParseFunctionReturn {
    const parsedObject = {};
    let position = pos;
    const definition = this.definitions[typeDefinitionName];
    for (const key in definition) {
      const value = definition[key];
      let parsedResult: ParseFunctionReturn;
      if (typeof value === "string") {
        if (this.DEBUG) console.log(">", key, position.toString(16), "(parseType)");
        parsedResult = this.parseType(dv, position, value);
      } else {
        const { baseType, subType, length } = value;
        if (this.DEBUG) console.log(">", key, position.toString(16), baseType, subType, length);
        parsedResult = this[baseType](dv, position, subType, length);
      }
      parsedObject[key] = parsedResult.data;

      position = parsedResult.newPosition;
    }

    return {
      newPosition: position,
      data: parsedObject,
    };
  }

  /**
   *      Parsers
   **/

  private Float32(dv: DataView, pos: number): ParseFunctionReturn {
    if (this.DEBUG) console.debug("Float32", dv.getFloat32(pos, true));
    return { newPosition: pos + 4, data: dv.getFloat32(pos, true) };
  }

  private Float64(dv: DataView, pos: number): ParseFunctionReturn {
    return { newPosition: pos + 8, data: dv.getFloat64(pos, true) };
  }

  private Uint8(dv: DataView, pos: number): ParseFunctionReturn {
    return { newPosition: pos + 1, data: dv.getUint8(pos) };
  }

  private Uint16(dv: DataView, pos: number): ParseFunctionReturn {
    return { newPosition: pos + 2, data: dv.getUint16(pos, true) };
  }

  private Uint32(dv: DataView, pos: number): ParseFunctionReturn {
    if (this.DEBUG) console.debug("Uint32", dv.getUint32(pos, true));
    return { newPosition: pos + 4, data: dv.getUint32(pos, true) };
  }

  private Uint64(dv: DataView, pos: number): ParseFunctionReturn {
    if (this.DEBUG) console.debug("Uint64", dv.getUint32(pos, true), dv.getUint32(pos + 4, true));
    return {
      newPosition: pos + 8,
      data: (BigInt(dv.getUint32(pos + 4, true)) << BigInt(32)) | BigInt(dv.getUint32(pos, true)),
    };
  }

  private CString(dv: DataView, pos: number, _subType: DataType | string, length: number): ParseFunctionReturn {
    const u8 = new Uint8Array(dv.buffer, pos);
    const end = length || u8.findIndex((v) => v === 0);

    if (this.DEBUG) console.debug("CString", String.fromCharCode.apply(null, new Uint8Array(u8.slice(0, end))));
    return {
      newPosition: pos + end,
      data: String.fromCharCode.apply(null, new Uint8Array(u8.slice(0, end))),
    };
  }

  private RefString(dv: DataView, pos: number): ParseFunctionReturn {
    let ptr = pos + dv.getUint32(pos, true);

    let data = "";
    let num;
    while (ptr + 1 < dv.byteLength && (num = dv.getUint8(ptr)) !== 0) {
      ptr += 1;
      data += String.fromCharCode(num);
    }

    return {
      newPosition: pos + 4,
      data,
    };
  }

  private FixedArray(dv: DataView, pos: number, type: DataType | string, length: number): ParseFunctionReturn {
    // Some types can be mapped directly from their buffer into the return type
    if (typeof type != "string" && this._getOptimisedArrayConstructor(type.baseType)) {
      return this.optimisedArray(dv, pos, type, length);
    }

    const data = [];
    let newPosition = pos;
    for (let itemIndex = 0; itemIndex < length; itemIndex++) {
      const parsedItem =
        typeof type === "string"
          ? this.parseType(dv, newPosition, type)
          : this[type.baseType](dv, newPosition, type.subType, type.length);
      data.push(parsedItem.data);
      newPosition = parsedItem.newPosition;
    }

    return {
      newPosition,
      data,
    };
  }

  private DynArray(dv: DataView, pos: number, type: DataType | string): ParseFunctionReturn {
    let arrayLength = dv.getUint32(pos, true);
    let arrayOffset = dv.getUint32(pos + 4, true);
    if (this.DEBUG) console.debug("DynArray", arrayLength, arrayOffset);

    if (arrayOffset === 0) {
      return {
        newPosition: pos + 8,
        data: [],
      };
    }
    let arrayPtr = pos + 4 + arrayOffset;

    return {
      newPosition: pos + 8,
      data: this.FixedArray(dv, arrayPtr, type, arrayLength).data,
    };
  }

  private RefArray(dv: DataView, pos: number, type: DataType | string): ParseFunctionReturn {
    let data = [];
    let arrayLength = dv.getUint32(pos, true);
    let arrayPtr = dv.getUint32(pos + 4, true) + pos + 4;
    if (this.DEBUG) console.debug("RefArray", "arrayPtr", (pos + 4).toString(16), arrayPtr.toString(16));
    if (arrayLength === 0) {
      return {
        data,
        newPosition: pos + 8,
      };
    }

    const pointer = dv.getUint32(pos + 4, true) + pos + 4;
    for (let index = 0; index < arrayLength; index++) {
      const offset = dv.getUint32(arrayPtr + 4 * index, true);
      if (offset !== 0) {
        if (this.DEBUG) console.debug("RefArray", "offset", offset.toString(16));
        let newPosition = pointer + index * 4 + offset;
        if (typeof type === "string") {
          data.push(this.parseType(dv, newPosition, type).data);
        } else {
          data.push(this[type.baseType](dv, newPosition, type.subType, type.length).data);
        }
      }
    }

    return {
      newPosition: pos + 8,
      data,
    };
  }

  private Pointer(dv: DataView, pos: number, type: DataType | string): ParseFunctionReturn {
    const offset = dv.getUint32(pos, true);
    const parsedItem =
      typeof type === "string"
        ? this.parseType(dv, pos + offset, type)
        : this[type.baseType](dv, pos + offset, type.subType, type.length);

    return {
      newPosition: pos + 4,
      data: parsedItem.data,
    };
  }

  private RefString16(dv: DataView, pos: number): ParseFunctionReturn {
    let ptr = pos + dv.getUint32(pos, true);

    let data = "";
    let num;
    while (ptr + 2 < dv.byteLength && (num = dv.getUint16(ptr, true)) !== 0) {
      ptr += 2;
      data += String.fromCharCode(num);
    }

    return {
      newPosition: pos + 4,
      data,
    };
  }

  private Fileref(dv: DataView, pos: number): ParseFunctionReturn {
    return this.Filename(dv, pos);
  }

  private Filename(dv: DataView, pos: number): ParseFunctionReturn {
    // This implementation is based on the old Utils.getFileNameReader() function
    let newPosition = pos;
    //try {
    let ptr = newPosition + dv.getUint32(newPosition, true);
    newPosition += 4;

    const m_lowPart = dv.getUint16(ptr, true);
    ptr += 2;
    const m_highPart = dv.getUint16(ptr, true);
    ptr += 2;
    const _m_terminator = dv.getUint16(ptr, true);
    ptr += 2;

    /// Getting the file name...
    /// Both need to be >= than 256 (terminator is 0)
    const ret = 0xff00 * (m_highPart - 0x100) + (m_lowPart - 0x100) + 1;

    return {
      newPosition: pos + 4,
      data: ret > 0 ? ret : 0,
    };
    // } catch (e) {
    //   return {
    //     newPosition,
    //     data: -1,
    //   };
    // }
  }

  private Unknown(dv: DataView, pos: number): ParseFunctionReturn {
    throw new Error("Could not parse unknown data");
  }

  /**
   *      Parser utils & helpers
   **/

  private optimisedArray(dv: DataView, pos: number, type: DataType, length: number): ParseFunctionReturn {
    const OptimisedArray = this._getOptimisedArrayConstructor(type.baseType);
    if (this.DEBUG) console.debug("OptimizedArray", Array.from(new OptimisedArray(dv.buffer, pos, length)));
    return {
      newPosition: pos + length * OptimisedArray.BYTES_PER_ELEMENT,
      data: Array.from(new OptimisedArray(dv.buffer, pos, length)),
    };
  }

  private _getOptimisedArrayConstructor(baseType: BaseType) {
    switch (baseType) {
      case BaseType.Float32:
        return Float32Array;

      case BaseType.Float64:
        return Float64Array;

      case BaseType.Uint8:
        return Uint8Array;

      case BaseType.Uint16:
        return Uint16Array;

      // case BaseType.Uint32:
      //   return Uint32Array;
    }
  }
}
