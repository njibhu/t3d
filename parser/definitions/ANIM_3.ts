import { Uint64, Fileref, Float32, Uint32, DynArray } from "../src/types";

export const V0 = {
  chunkName: "anim",
  name: "PackEmoteAnimationsV0",
  version: 0,
  definitions: {
    PackEmoteAnimationV0: {
      token: Uint64,
      timings: DynArray("PackEmoteTimingV0")
    },
    PackEmoteTimingV0: {
      modelFileId: Fileref(),
      blendInTime: Float32,
      blendOutTime: Float32,
      duration: Uint32
    }
  },
  root: {
    animations: DynArray("PackEmoteAnimationV0")
  }
};

export const V1 = {
  chunkName: "anim",
  name: "PackEmoteAnimationsV1",
  version: 1,
  definitions: {
    PackEmoteAnimationV1: {
      token: Uint64,
      timings: DynArray("PackEmoteTimingV1")
    },
    PackEmoteTimingV1: {
      modelFileId: Fileref(),
      blendInTime: Float32,
      blendOutTime: Float32,
      duration: Uint32,
      loopDuration: Uint32
    }
  },
  root: {
    animations: DynArray("PackEmoteAnimationV1")
  }
};

export const V2 = {
  chunkName: "anim",
  name: "PackEmoteAnimationsV2",
  version: 2,
  definitions: {
    PackEmoteAnimationV2: {
      Token: Uint64,
      Timing: DynArray("PackEmoteTimingV2")
    },
    PackEmoteTimingV2: {
      ModelFile: Fileref(),
      BlendIn: Float32,
      BlendOut: Float32,
      IntroDuration: Uint32,
      LoopDuration: Uint32,
      OutroDuration: Uint32
    }
  },
  root: {
    Animation: DynArray("PackEmoteAnimationV2")
  }
};

export const V3 = {
  chunkName: "anim",
  name: "PackEmoteAnimationsV3",
  version: 3,
  definitions: {
    PackEmoteAnimationV3: {
      Token: Uint64,
      Timing: DynArray("PackEmoteTimingV3")
    },
    PackEmoteTimingV3: {
      ModelFile: Fileref(),
      BlendIn: Float32,
      BlendOut: Float32,
      IntroDuration: Uint32,
      LoopDuration: Uint32,
      OutroDuration: Uint32,
      StartOffset: Uint32
    }
  },
  root: {
    Animation: DynArray("PackEmoteAnimationV3")
  }
};

export const latest = V3;
export const definitionArray = [V0, V1, V2, V3];