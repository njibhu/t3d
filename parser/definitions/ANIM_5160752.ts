import { Uint64, Float32, FixedArray, DynArray, Filename, Uint32 } from "../src/types";

export const V0 = {
  chunkName: "ANIM",
  name: "SceneFileAnimationV0",
  version: 0,
  definitions: {
    SceneAnimationV0: {
      name: Uint64,
      motion: "SceneMotionV0",
      actionPoints: DynArray("SceneActionPointV0")
    },
    SceneMotionV0: {
      keys: DynArray("SceneKeyframeV0")
    },
    SceneKeyframeV0: {
      time: Float32,
      position: FixedArray(Float32, 3),
      rotation: FixedArray(Float32, 4)
    },
    SceneActionPointV0: {
      name: Uint64,
      motion: "SceneMotionV0"
    },
    SceneAnimationImportV0: {
      filename: Filename(),
      animNames: DynArray(Uint64)
    }
  },
  root: {
    animations: DynArray("SceneAnimationV0"),
    imports: DynArray("SceneAnimationImportV0")
  }
};

export const V1 = {
  chunkName: "ANIM",
  name: "SceneFileAnimationV1",
  version: 1,
  definitions: {
    SceneAnimationV1: {
      name: Uint64,
      motion: "SceneMotionV1",
      actionPoints: DynArray("SceneActionPointV1"),
      events: DynArray("SceneAnimationEventV1")
    },
    SceneMotionV1: {
      keys: DynArray("SceneKeyframeV1")
    },
    SceneKeyframeV1: {
      time: Float32,
      position: FixedArray(Float32, 3),
      rotation: FixedArray(Float32, 4)
    },
    SceneActionPointV1: {
      name: Uint64,
      motion: "SceneMotionV1"
    },
    SceneAnimationEventV1: {
      name: Uint64,
      time: Float32
    },
    SceneAnimationImportV1: {
      filename: Filename(),
      animNames: DynArray(Uint64)
    }
  },
  root: {
    animations: DynArray("SceneAnimationV1"),
    imports: DynArray("SceneAnimationImportV1")
  }
};

export const V2 = {
  chunkName: "ANIM",
  name: "SceneFileAnimationV2",
  version: 2,
  definitions: {
    SceneAnimationV2: {
      name: Uint64,
      motion: "SceneMotionV2",
      actionPoints: DynArray("SceneActionPointV2"),
      events: DynArray("SceneAnimationEventV2")
    },
    SceneMotionV2: {
      keys: DynArray("SceneKeyframeV2")
    },
    SceneKeyframeV2: {
      time: Float32,
      position: FixedArray(Float32, 3),
      rotation: FixedArray(Float32, 4)
    },
    SceneActionPointV2: {
      name: Uint64,
      motion: "SceneMotionV2"
    },
    SceneAnimationEventV2: {
      name: Uint64,
      time: Float32
    },
    ScenePoseV2: {
      name: Uint64,
      transforms: DynArray("SceneTransformV2")
    },
    SceneTransformV2: {
      name: Uint64,
      translation: FixedArray(Float32, 3),
      rotation: FixedArray(Float32, 4)
    },
    SceneAnimationImportV2: {
      filename: Filename(),
      animNames: DynArray("SceneImportSequenceV2"),
      flags: Uint32
    },
    SceneImportSequenceV2: {
      name: Uint64
    }
  },
  root: {
    animations: DynArray("SceneAnimationV2"),
    poses: DynArray("ScenePoseV2"),
    imports: DynArray("SceneAnimationImportV2")
  }
};

export const latest = V2;
export const definitionArray = [V0, V1, V2];