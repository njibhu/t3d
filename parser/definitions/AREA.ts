import { Uint64, Uint8, Uint32, FixedArray, DynArray, Float32, RefString16 } from "../src/types";

export const V0 = {
  chunkName: "area",
  name: "PackMapAreasV0",
  version: 0,
  definitions: {
    PackMapAreaV0: {
      min: FixedArray(Float32, 3),
      max: FixedArray(Float32, 3),
      internal: FixedArray(Float32, 3),
      external: FixedArray(Float32, 3),
      token: Uint32,
      flags: Uint32
    },
    PackMapPortalV0: {
      portalVerts: DynArray(FixedArray(Float32, 3))
    }
  },
  root: {
    areas: DynArray("PackMapAreaV0"),
    portals: DynArray("PackMapPortalV0")
  }
};

export const V1 = {
  chunkName: "area",
  name: "PackMapAreasV1",
  version: 1,
  definitions: {
    PackMapAreaV1: {
      token: Uint64,
      type: Uint8,
      floor: Uint8,
      polygon: "PackMapAreaPolygonV1",
      volume: "PackMapAreaVolumeV1"
    },
    PackMapAreaPolygonV1: {
      points: DynArray(FixedArray(Float32, 3)),
      height: Float32
    },
    PackMapAreaVolumeV1: {
      portals: DynArray("PackMapAreaPortalV1"),
      position: FixedArray(Float32, 3),
      extents: FixedArray(Float32, 3),
      pointInterior: FixedArray(Float32, 3),
      pointExterior: FixedArray(Float32, 3)
    },
    PackMapAreaPortalV1: {
      position: FixedArray(Float32, 3),
      extents: FixedArray(Float32, 3),
      rotation: FixedArray(Float32, 3)
    }
  },
  root: {
    areas: DynArray("PackMapAreaV1")
  }
};

export const V2 = {
  chunkName: "area",
  name: "PackMapAreasV2",
  version: 2,
  definitions: {
    PackMapAreaV2: {
      token: Uint64,
      type: Uint8,
      floor: Uint8,
      flags: Uint32,
      polygon: "PackMapAreaPolygonV2",
      volume: "PackMapAreaVolumeV2"
    },
    PackMapAreaPolygonV2: {
      points: DynArray(FixedArray(Float32, 3)),
      height: Float32
    },
    PackMapAreaVolumeV2: {
      portals: DynArray("PackMapAreaPortalV2"),
      position: FixedArray(Float32, 3),
      extents: FixedArray(Float32, 3),
      pointInterior: FixedArray(Float32, 3),
      pointExterior: FixedArray(Float32, 3)
    },
    PackMapAreaPortalV2: {
      position: FixedArray(Float32, 3),
      extents: FixedArray(Float32, 3),
      rotation: FixedArray(Float32, 3)
    },
    PackMapAreaToolV2: {
      annotation: RefString16()
    }
  },
  root: {
    areas: DynArray("PackMapAreaV2"),
    areaTools: DynArray("PackMapAreaToolV2")
  }
};

export const V3 = {
  chunkName: "area",
  name: "PackMapAreasV3",
  version: 3,
  definitions: {
    PackMapAreaV3: {
      token: Uint64,
      type: Uint8,
      floor: Uint8,
      flags: Uint32,
      polygon: "PackMapAreaPolygonV3",
      volume: "PackMapAreaVolumeV3"
    },
    PackMapAreaPolygonV3: {
      points: DynArray(FixedArray(Float32, 3)),
      height: Float32
    },
    PackMapAreaVolumeV3: {
      portals: DynArray("PackMapAreaPortalV3"),
      position: FixedArray(Float32, 3),
      extents: FixedArray(Float32, 3),
      pointInterior: FixedArray(Float32, 3),
      pointExterior: FixedArray(Float32, 3)
    },
    PackMapAreaPortalV3: {
      position: FixedArray(Float32, 3),
      extents: FixedArray(Float32, 3),
      rotation: FixedArray(Float32, 3)
    },
    PackMapAreaToolV3: {
      annotation: RefString16(),
      renderOffset: Float32
    }
  },
  root: {
    areas: DynArray("PackMapAreaV3"),
    areaTools: DynArray("PackMapAreaToolV3")
  }
};

export const V4 = {
  chunkName: "area",
  name: "PackMapAreasV4",
  version: 4,
  definitions: {
    PackMapAreaV4: {
      token: Uint64,
      type: Uint8,
      floor: Uint8,
      flags: Uint32,
      polygon: "PackMapAreaPolygonV4",
      volume: "PackMapAreaVolumeV4"
    },
    PackMapAreaPolygonV4: {
      points: DynArray(FixedArray(Float32, 3)),
      height: Float32
    },
    PackMapAreaVolumeV4: {
      portals: DynArray("PackMapAreaPortalV4"),
      position: FixedArray(Float32, 3),
      extents: FixedArray(Float32, 3),
      pointInterior: FixedArray(Float32, 3),
      pointExterior: FixedArray(Float32, 3)
    },
    PackMapAreaPortalV4: {
      position: FixedArray(Float32, 3),
      extents: FixedArray(Float32, 3),
      rotation: FixedArray(Float32, 3)
    },
    PackMapAreaToolV4: {
      annotation: RefString16(),
      renderOffset: Float32
    }
  },
  root: {
    areas: DynArray("PackMapAreaV4"),
    areaTools: DynArray("PackMapAreaToolV4")
  }
};

export const latest = V4;
export const definitionArray = [V0, V1, V2, V3, V4];