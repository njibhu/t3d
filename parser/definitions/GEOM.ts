import { Uint64, Uint16, FixedArray, Float32, DynArray, Uint32, CString, Uint8, Pointer, RefArray } from "../src/types";

export const V0 = {
  chunkName: "GEOM",
  name: "ModelFileGeometryV0",
  version: 0,
  definitions: {
    ModelMeshDataV65: {
      visBone: Uint64,
      morphTargets: DynArray("ModelMeshMorphTargetV65"),
      flags: Uint32,
      seamVertIndices: DynArray(Uint32),
      meshName: Uint64,
      minBound: FixedArray(Float32, 3),
      maxBound: FixedArray(Float32, 3),
      bounds: DynArray("GrBoundData"),
      materialIndex: Uint32,
      materialName: CString(),
      boneNames: DynArray(CString()),
      geometry: Pointer("ModelMeshGeometryV0")
    },
    ModelMeshMorphTargetV65: {
      positions: DynArray("ModelMeshMorphVertV65"),
      normals: DynArray("ModelMeshMorphVertV65"),
      mesh: Uint64
    },
    ModelMeshMorphVertV65: {
      index: Uint16,
      vector: FixedArray(Float32, 3)
    },
    GrBoundData: {
      center: FixedArray(Float32, 3),
      boxExtent: FixedArray(Float32, 3),
      sphereRadius: Float32
    },
    ModelMeshGeometryV0: {
      verts: "ModelMeshVertexDataV0",
      indices: "ModelMeshIndexDataV0",
      lods: DynArray("ModelMeshIndexDataV0")
    },
    ModelMeshVertexDataV0: {
      vertexCount: Uint32,
      mesh: "PackVertexType"
    },
    PackVertexType: {
      fvf: Uint32,
      vertices: DynArray(Uint8)
    },
    ModelMeshIndexDataV0: {
      indices: DynArray(Uint16)
    }
  },
  root: {
    meshes: RefArray("ModelMeshDataV65")
  }
};

export const V1 = {
  chunkName: "GEOM",
  name: "ModelFileGeometryV1",
  version: 1,
  definitions: {
    ModelMeshDataV66: {
      visBone: Uint64,
      morphTargets: DynArray("ModelMeshMorphTargetV66"),
      flags: Uint32,
      seamVertIndices: DynArray(Uint32),
      meshName: Uint64,
      minBound: FixedArray(Float32, 3),
      maxBound: FixedArray(Float32, 3),
      bounds: DynArray("GrBoundData"),
      materialIndex: Uint32,
      materialName: CString(),
      boneBindings: DynArray(Uint64),
      geometry: Pointer("ModelMeshGeometryV1")
    },
    ModelMeshMorphTargetV66: {
      positions: DynArray("ModelMeshMorphVertV66"),
      normals: DynArray("ModelMeshMorphVertV66"),
      mesh: Uint64
    },
    ModelMeshMorphVertV66: {
      index: Uint16,
      vector: FixedArray(Float32, 3)
    },
    GrBoundData: {
      center: FixedArray(Float32, 3),
      boxExtent: FixedArray(Float32, 3),
      sphereRadius: Float32
    },
    ModelMeshGeometryV1: {
      verts: "ModelMeshVertexDataV1",
      indices: "ModelMeshIndexDataV1",
      lods: DynArray("ModelMeshIndexDataV1"),
      transforms: DynArray(Uint32)
    },
    ModelMeshVertexDataV1: {
      vertexCount: Uint32,
      mesh: "PackVertexType"
    },
    PackVertexType: {
      fvf: Uint32,
      vertices: DynArray(Uint8)
    },
    ModelMeshIndexDataV1: {
      indices: DynArray(Uint16)
    }
  },
  root: {
    meshes: RefArray("ModelMeshDataV66")
  }
};

export const latest = V1;
export const definitionArray = [V0, V1];