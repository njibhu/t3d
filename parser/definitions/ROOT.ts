import { Uint64, Float32, DynArray, FixedArray } from "../src/types";

export const V0 = {
  chunkName: "ROOT",
  name: "ModelFileRootMotionV0",
  version: 0,
  definitions: {
    ModelRootMotionV0: {
      sequence: Uint64,
      keys: DynArray(Float32),
      values: DynArray(FixedArray(Float32, 3))
    }
  },
  root: {
    rootMotions: DynArray("ModelRootMotionV0")
  }
};

export const V1 = {
  chunkName: "ROOT",
  name: "ModelFileRootMotionV1",
  version: 1,
  definitions: {
    ModelRootMotionV1: {
      sequence: Uint64,
      keys: DynArray(Float32),
      posValues: DynArray(FixedArray(Float32, 3)),
      quatValues: DynArray(FixedArray(Float32, 4))
    }
  },
  root: {
    rootMotions: DynArray("ModelRootMotionV1")
  }
};

export const latest = V1;
export const definitionArray = [V0, V1];