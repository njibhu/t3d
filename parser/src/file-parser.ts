import { CString, Uint16, Uint32 } from "../src/types";
import { DataParser } from "./data-parser";

interface FileHead {
  identifier: string;
  unknownField1: number;
  unknownField2: number;
  pkFileVersion: number;
  type: string;
}

const FILE_HEAD = {
  definitions: {},
  root: {
    identifier: CString(2),
    unknownField1: Uint16,
    unknownField2: Uint16,
    pkFileVersion: Uint16,
    type: CString(4),
  },
};

export interface ChunkHead {
  type: string;
  chunkDataSize: number;
  chunkVersion: number;
  chunkHeaderSize: number;
  offsetTableOffset: number;
}

const CHUNK_HEAD = {
  definitions: {},
  root: {
    type: CString(4),
    chunkDataSize: Uint32,
    chunkVersion: Uint16,
    chunkHeaderSize: Uint16,
    offsetTableOffset: Uint32,
  },
};

export function parseFile(dataView: DataView): { newPosition: number; data: FileHead } {
  const fileHeaderParser = new DataParser(FILE_HEAD);
  return fileHeaderParser.parse(dataView, 0);
}

export function parseChunkHead(dataView: DataView, pos: number): { newPosition: number; data: ChunkHead } {
  const chunkHeadParser = new DataParser(CHUNK_HEAD);
  return chunkHeadParser.parse(dataView, pos);
}

export function parseAllChunks(
  dataView: DataView,
  firstChunkPosition: number
): { chunkPosition: number; chunkHeader: ChunkHead }[] {
  const chunks = [];
  let pos = firstChunkPosition;
  while (pos < dataView.byteLength) {
    const chunk = parseChunkHead(dataView, pos);
    chunks.push({ chunkPosition: pos, chunkHeader: chunk.data });
    pos = chunk.newPosition - 8 + chunk.data.chunkDataSize;
  }
  return chunks;
}
