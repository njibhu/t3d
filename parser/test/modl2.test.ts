import * as fs from "fs";

import { DataParser } from "../src/data-parser";
import { parseFile, parseAllChunks } from "../src/file-parser";

import {  toArrayBuffer } from "./test-helper";

import * as MODL from "../definitions/MODL";
import * as PRPS from "../definitions/PRPS";
import * as ANIM from "../definitions/ANIM";
import * as GEOM from "../definitions/GEOM";
import * as ROOT from "../definitions/ROOT";

const chunkBuffer = fs.readFileSync("./test/modl2.bin", null);
const dv = new DataView(toArrayBuffer(chunkBuffer));
const fileHead = parseFile(dv);
const allChunks = parseAllChunks(dv, fileHead.newPosition);
const is64Bit = fileHead.data.flags == 5;

describe("modl2", () => {
  test("matches for MODL", function () {
    const modlChunk = allChunks.find((c) => c.chunkHeader.type === "MODL");
    const def = MODL.definitions[`V${modlChunk!.chunkHeader.chunkVersion}` as keyof typeof MODL["definitions"]];
    const test = new DataParser(def, is64Bit).parse(dv, modlChunk!.chunkPosition + modlChunk!.chunkHeader.chunkHeaderSize);
    expect(test.data).toMatchSnapshot("modl2-modl")
  });

  test("matches for PRPS", function () {
    const prpsChunk = allChunks.find((c) => c.chunkHeader.type === "PRPS");
    const def = PRPS.definitions[`V${prpsChunk!.chunkHeader.chunkVersion}` as keyof typeof PRPS["definitions"]];
    const test = new DataParser(def, is64Bit).parse(dv, prpsChunk!.chunkPosition + prpsChunk!.chunkHeader.chunkHeaderSize);
    expect(test.data).toMatchSnapshot("modl2-prps")
  });

  test("matches for ANIM", function () {
    const animChunk = allChunks.find((c) => c.chunkHeader.type === "ANIM");
    const def = ANIM.definitions[`V${animChunk!.chunkHeader.chunkVersion}` as keyof typeof ANIM["definitions"]];
    const test = new DataParser(def, is64Bit).parse(dv, animChunk!.chunkPosition + animChunk!.chunkHeader.chunkHeaderSize);
    expect(test.data).toMatchSnapshot("modl2-anim")
  });

  // GEOM
  test("matches for GEOM", function () {
    const geomChunk = allChunks.find((c) => c.chunkHeader.type === "GEOM");
    const def = GEOM.definitions[`V${geomChunk!.chunkHeader.chunkVersion}` as keyof typeof GEOM["definitions"]];
    const parser = new DataParser(def, is64Bit);
    // parser.DEBUG = true;
    const test = parser.parse(dv, geomChunk!.chunkPosition + geomChunk!.chunkHeader.chunkHeaderSize);

    expect(test.data).toMatchSnapshot("modl2-geom")
  });

  //ROOT
  test("matches for ROOT", function () {
    const rootChunk = allChunks.find((c) => c.chunkHeader.type === "ROOT");
    const def = ROOT.definitions[`V${rootChunk!.chunkHeader.chunkVersion}` as keyof typeof ROOT["definitions"]];
    const parser = new DataParser(def, is64Bit);
    //parser.DEBUG = true;
    const test = parser.parse(dv, rootChunk!.chunkPosition + rootChunk!.chunkHeader.chunkHeaderSize);
    expect(test.data).toMatchSnapshot("modl2-root")
  });
});
