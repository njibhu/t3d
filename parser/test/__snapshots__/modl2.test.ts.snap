// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modl2 matches for ANIM: modl2-anim 1`] = `
Object {
  "anim": Object {
    "animation": Array [],
    "pointers": Array [],
  },
  "bank": Object {
    "animations": Array [],
    "compoundAnimations": Array [],
    "fallbacks": Array [],
    "imports": Array [],
  },
}
`;

exports[`modl2 matches for GEOM: modl2-geom 1`] = `
{
  "meshes": [
    {
      "boneBindings": [],
      "bounds": [],
      "flags": 0,
      "geometry": {},
      "materialIndex": 4,
      "materialName": "",
      "maxBound": Float32Array [
        -33.14767837524414,
        -128.37332153320312,
        13.069268226623535,
      ],
      "meshName": 84n,
      "minBound": Float32Array [
        3.5342708218127104e-19,
        1.977759021384903e-39,
        -13.069268226623535,
      ],
      "morphTargets": [],
      "seamVertIndices": Uint32Array [],
      "visBone": 0n,
    },
  ],
}
`;

exports[`modl2 matches for MODL: modl2-modl 1`] = `
Object {
  "actionOffsetNames": Array [],
  "actionOffsets": Array [],
  "boneOffsetData": Array [],
  "boundingSphere": Object {},
  "clothData": Array [],
  "cloudData": Object {},
  "lightData": Object {},
  "lightningData": Object {},
  "lodOverride": Float32Array [
    0,
    0,
  ],
  "obstacles": Array [],
  "permutations": Array [
    Object {
      "materials": Array [
        Object {
          "constants": Array [],
          "filename": 2313158,
          "matConstLinks": Array [
            Object {
              "constantToken": 1061427347,
              "linkToken": 4458347457727686369n,
            },
            Object {
              "constantToken": 0,
              "linkToken": 0n,
            },
            Object {
              "constantToken": 1047233823,
              "linkToken": 4424048046607020758n,
            },
            Object {
              "constantToken": 2,
              "linkToken": 0n,
            },
            Object {
              "constantToken": 1047233823,
              "linkToken": 4453159315984593812n,
            },
            Object {
              "constantToken": 2,
              "linkToken": 0n,
            },
            Object {
              "constantToken": 1065353216,
              "linkToken": 4575657222285340212n,
            },
          ],
          "materialFlags": 0,
          "materialId": 9,
          "sortLayer": 0,
          "sortOrder": 2568,
          "texCoordCount": 0,
          "texTransforms": Array [],
          "textures": Array [],
          "token": 19379777n,
          "uvTransLinks": Array [],
        },
      ],
      "token": 0n,
    },
  ],
  "softBodyData": Array [],
  "soundScript": 0,
  "streakData": Object {},
  "windData": Object {},
}
`;

exports[`modl2 matches for PRPS: modl2-prps 1`] = `
Object {
  "fixedOffsetData": Array [],
  "properties": Array [],
}
`;

exports[`modl2 matches for ROOT: modl2-root 1`] = `
Object {
  "rootMotions": Array [],
}
`;
